### PR TITLE
Fix getDispatchEnv / getDamlEnv re-entrancy.

### PR DESCRIPTION
**Pull request by associahedron**
Short version: I fixed some bugs that would cause daml assistant to compute its environment wrongly.

Long version: While working on #880 I noticed that the `damlEnvVars` definition was out of date, so I wrote a test that would catch this (that `getDispatchEnv` should be idempotent, i.e. when applied twice to an environment it will return the same environment as when applied once). But there was another bug that I noticed -- that getDamlEnv wouldn't pick up the environment variables. That is fixed now, but it also involved tracking down some unfortunate behavior in `withEnv` as a result of `System.Environment.setEnv` incorrect handling of empty strings, that was affecting the correctness of our tests.

### CODE REVIEW

File: .hlint.yaml [hunk 1]
```diff
@@ -45,6 +45,7 @@
 - warn: {lhs: Data.Text.writeFile, rhs: Data.Text.Extended.writeFileUtf8}
 - warn: {lhs: Data.Text.Lazy.readFile, rhs: Data.Text.Extended.readFileUtf8}
 - warn: {lhs: Data.Text.Lazy.writeFile, rhs: Data.Text.Extended.writeFileUtf8}
+- warn: {lhs: System.Environment.setEnv, rhs: System.Environment.Blank.setEnv}
```
Comment (cocreature) [inline .hlint.yaml]: :+1:

File: daml-assistant/BUILD.bazel [hunk 2]
```diff
@@ -48,7 +48,6 @@ da_haskell_library(
         "http-client-tls",
         "http-conduit",
         "lens",
-        "main-tester",
```
Comment (cocreature) [inline daml-assistant/BUILD.bazel]: :fire:

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 3]
```diff
@@ -59,11 +59,51 @@ getDamlEnv = do
...
 
+-- | (internal) Override function with environment variable
+-- if it is available.
+overrideWithEnvVar
+    :: String                   -- ^ env var name
+    -> (String -> t)            -- ^ parser for env var
+    -> IO t                     -- ^ calculation to override
+    -> IO t
+overrideWithEnvVar envVar parse calculate = do
+    valueM <- getEnv envVar
```
Comment (cocreature) [inline daml-assistant/src/DAML/Assistant/Env.hs]:
```suggestion
    maybeM calculate parse (getEnv envVar)
```

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 4]
```diff
@@ -59,11 +59,51 @@ getDamlEnv = do
...
+
+-- | (internal) Same as overrideWithEnvVar but accepts "" as
+-- Nothing and throws exception on parse failure.
+overrideWithEnvVarMaybe
+    :: Exception e
+    => String                   -- ^ env var name
+    -> (String -> Either e t)   -- ^ parser for env var
+    -> IO (Maybe t)             -- ^ calculation to override
+    -> IO (Maybe t)
+overrideWithEnvVarMaybe envVar parse calculate = do
```
Comment (cocreature) [inline daml-assistant/src/DAML/Assistant/Env.hs]: I feel like we should be able to simplify this somehow but I also don’t have a good suggestion.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 5]
```diff
@@ -90,18 +130,40 @@ testDamlEnv Env{..} = firstJustM (\(test, msg) -> unlessMaybeM test (pure msg))
...
     | takeFileName damlPath == ".daml-head" = pure $ DamlAssistantPath $ damlPath </> "bin" </> "daml-head"
     | otherwise = pure $ DamlAssistantPath $ damlPath </> "bin" </> "daml"
 
--- | Determine SDK version of running daml assistant.
+-- | Determine SDK version of running daml assistant. Can be overriden
+-- with DAML_ASSISTANT_VERSION env var.
 getDamlAssistantSdkVersion :: IO (Maybe DamlAssistantSdkVersion)
-getDamlAssistantSdkVersion = fmap DamlAssistantSdkVersion <$> do
+getDamlAssistantSdkVersion = do
+    versionStrM <- getEnv damlAssistantVersionEnvVar
```
Comment (cocreature) [inline daml-assistant/src/DAML/Assistant/Env.hs]: That looks like you are replicating `overrideWithEnvVarMaybe`.

File: daml-assistant/test/DAML/Assistant/Tests.hs [hunk 6]
```diff
@@ -36,14 +35,39 @@ import qualified Data.Conduit.Tar as Tar
 -- unix specific
 import System.PosixCompat.Files (createSymbolicLink)
 
+-- | Replace all environment variables for test action, then restore them.
```
Comment (cocreature) [inline daml-assistant/test/DAML/Assistant/Tests.hs]: A comment mentioning the `setEnv` behavior might be useful.

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 5 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Env.hs]: oops!

File: daml-assistant/test/DAML/Assistant/Tests.hs (hunk 6 above)
Comment (associahedron) [reply daml-assistant/test/DAML/Assistant/Tests.hs]: :+1:

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 7]
```diff
@@ -59,11 +59,48 @@ getDamlEnv = do
...
+overrideWithEnvVar envVar parse calculate =
+    maybeM calculate (pure . parse) (getEnv envVar)
+
+-- | (internal) Same as overrideWithEnvVar but accepts "" as
+-- Nothing and throws exception on parse failure.
+overrideWithEnvVarMaybe
+    :: Exception e
+    => String                   -- ^ env var name
+    -> (String -> Either e t)   -- ^ parser for env var
+    -> IO (Maybe t)             -- ^ calculation to override
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: Is this an override or a default? I don't get how this overrides?

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 8]
```diff
@@ -147,10 +199,9 @@ getDamlPath = wrapErr "Determining daml home directory." $ do
 -- environment variable.
 getProjectPath :: IO (Maybe ProjectPath)
 getProjectPath = wrapErr "Detecting daml project." $ do
-        pathM <- firstJustM id
-            [ lookupEnv projectPathEnvVar
-            , findM hasProjectConfig . ascendants =<< getCurrentDirectory
-            ]
+        pathM <- overrideWithEnvVarMaybe @SomeException projectPathEnvVar Right $ do
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: Should this be `@Void` and we have an exception instance for void? Or a `NoException` void like type? I hate seeing SomeException in the code.

**Pull request merged**